sql/core/src/test/scala/org/apache/spark/sql/UpdateFieldsBenchmark.scala

```diff
@@ -0,0 +1,310 @@
+  }
+
+  // simulates how a user would add/drop nested fields in a performant manner
+  def modifyNestedColumns(
+      column: Column,
+      numsToAdd: Seq[Int],
+      numsToDrop: Seq[Int],
+      maxDepth: Int,
+      currDepth: Int = 1): Column = {
+
+    // drop columns at the current depth
+    val dropped = if (numsToDrop.nonEmpty) {
+      column.dropFields(numsToDrop.map(num => nestedColName(currDepth, num)): _*)
+    } else column
+
+    // add columns at the current depth
+    val added = numsToAdd.foldLeft(dropped) {
+      (res, num) => res.withField(nestedColName(currDepth, num), lit(num))
+    }
+
+    if (currDepth == maxDepth) {
+      added
+    } else {
+      // add/drop columns at the next depth
+      val newValue = modifyNestedColumns(
+        column = col((0 to currDepth).map(d => nestedColName(d, 0)).mkString(".")),
+        numsToAdd = numsToAdd,
+        numsToDrop = numsToDrop,
+        currDepth = currDepth + 1,
+        maxDepth = maxDepth)
+      added.withField(nestedColName(currDepth, 0), newValue)
+    }
+  }
+
+  def updateFieldsBenchmark(
+      maxDepth: Int,
+      initialNumberOfColumns: Int,
+      numsToAdd: Seq[Int] = Seq.empty,
+      numsToDrop: Seq[Int] = Seq.empty): Unit = {
+
+    val name = s"Add ${numsToAdd.length} columns and drop ${numsToDrop.length} columns " +
+      s"at $maxDepth different depths of nesting"
+
+    runBenchmark(name) {
+      val benchmark = new Benchmark(
+        name = name,
+        // Because the point of this benchmark is only to ensure Spark is able to analyze and
+        // optimize long UpdateFields chains quickly, this benchmark operates only over 1 row of
+        // data.
+        valuesPerIteration = 1,
+        output = output)
+
+      val columnFunc = modifyNestedColumns(
+        col(nestedColName(0, 0)),
+        numsToAdd,
+        numsToDrop,
+        maxDepth
+      ).as(nestedColName(0, 0))
+
+      val nonNullableInputDf = nestedDf(maxDepth, initialNumberOfColumns, nullable = false)
+      val nullableInputDf = nestedDf(maxDepth, initialNumberOfColumns, nullable = true)
+
+      benchmark.addCase("Non-Nullable StructTypes") { _ =>
+        nonNullableInputDf.select(columnFunc).noop()
+      }
+
+      benchmark.addCase("Nullable StructTypes") { _ =>
+        nullableInputDf.select(columnFunc).noop()
+      }
+
+      benchmark.run()
+    }
+  }
+
+  override def runBenchmarkSuite(mainArgs: Array[String]): Unit = {
+    val maxDepth = 20
+
+    updateFieldsBenchmark(
+      maxDepth = maxDepth,
+      initialNumberOfColumns = 5,
+      numsToAdd = 5 to 9)
+
+    updateFieldsBenchmark(
+      maxDepth = maxDepth,
+      initialNumberOfColumns = 10,
+      numsToDrop = 5 to 9)
+
+    updateFieldsBenchmark(
+      maxDepth = maxDepth,
+      initialNumberOfColumns = 10,
+      numsToAdd = 10 to 14,
+      numsToDrop = 5 to 9)
+  }
+}
+
+class UpdateFieldsBenchmark extends QueryTest with SharedSparkSession {
```

Comment:
do we still need it?

Reply:
I removed some unnecessary tests but I would prefer to keep the rest as the methods being tested are not entirely trivial (e.g. recursion is involved).

Reply:
Then let's turn it into test, not benchmark. Can we move the tests to `ColumnExpressionSuite`?

Reply:
Sure, done.